### PR TITLE
Refactor InformSubscriber, missed notification for emails

### LINKED ISSUE #295: Failed MAIL notifications are not saved to database
### Description
When a subscription is triggered and MAIL is chosen as the notification type, any missed notification will not be stored to the database if Eiffel Intelligence was unable to send the email to the subscriber. No missed notification will be created and saved in the database, and the user will not be informed of this.
 
It should also be noted that EI assumes the configuration for "email.sender" and "email.subject" is set in application.properties, but there is no check on this upon startup. It is possible for users to create subscriptions with MAIL notifications but if EI is not configured with email.sender or email.subject then the notification will fail - and the user will never know.

The documentation should also be updated if needed.

### Motivation
A missed notification should be created also after a failed MAIL notification (not only REST POST). 

### Exemplification
If the user has not specified the right host/port of the mail server, then EI will assume a default mail server running on localhost and port 25. When this fails (in most cases) EI will log an exception, but not save the attempted notification to the database. This means the user will never know that their subscription was triggered, but keep waiting for that email. 

### Benefits
More reliable behavior by EI, and MAIL notifications will also be saved in missed notification database.

### Possible Drawbacks
None that I can think of.

### PR DESCRIPTION
**Pull request by a-breid-ericsson**
### Applicable Issues
InformSubscriber.java was large and hard to understand.
Missed notifications was not saved for emails. See issue #295 

### Description of the Change
InformSubscriber split to many classes.
New classes HttpRequest, JenkinsCrumb, UrlParser
Missed notifications is saved for both emails and rest post calls. Fixes #295 

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

Signed-off-by:

### CODE REVIEW

File: src/main/java/com/ericsson/ei/notifications/HttpRequest.java [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+     * @param password
+     * @return
+     */
+    private boolean isAuthenticationDetailsProvided(String authType, String username,
+            String password) {
+        if (authType.isEmpty() || authType.equals(AUTHENTICATION_TYPE_NO_AUTH)) {
+            return false;
+        }
+
+        if (username.equals("") && password.equals("")) {
```
[the remark this answers is not in the document]
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/notifications/HttpRequest.java]: They can never be null, the SubscriptionField.get returns empty string if value is not found.

[the remark this answers is not in the document]
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/notifications/HttpRequest.java]: Although the check should be .isEmpty()

File: src/main/java/com/ericsson/ei/notifications/HttpRequestSender.java [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            return false;
+        }
+
+        HttpStatus status = response.getStatusCode();
+        boolean httpStatusSuccess = status == HttpStatus.OK || status == HttpStatus.ACCEPTED
+                || status == HttpStatus.CREATED;
+
+        JsonNode body = response.getBody();
+
+        LOGGER.debug("The HTTP post request response status code is [{}] and body: {}", status, body);
```
[the remark this answers is not in the document]
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/notifications/HttpRequestSender.java]: I prefer [] around numerics and '' around text.

File: src/functionaltests/java/com/ericsson/ei/notifications/trigger/SubscriptionNotificationSteps.java [hunk 3]
```diff
@@ -69,26 +73,46 @@
...
     private ClientAndServer restServer;
     private MockServerClient mockClient;
     private ResponseEntity response;
 
-    @Before("@SubscriptionTriggerScenario")
-    public void beforeScenario() throws IOException {
-        setupSMTPServer();
+    @Before()
+    public void beforeScenario(Scenario scenario) throws IOException {
+        System.out.println("Setting up REST endpoints");
```
[the remark this answers is not in the document]
Comment (a-breid-ericsson) [reply src/functionaltests/java/com/ericsson/ei/notifications/trigger/SubscriptionNotificationSteps.java]: Already done.

File: src/functionaltests/java/com/ericsson/ei/notifications/trigger/SubscriptionNotificationSteps.java [hunk 4]
```diff
@@ -69,26 +73,46 @@
...
         setupRestEndpoints();
     }
 
-    @After("@SubscriptionTriggerScenario")
+    @Before("@SubscriptionNotificationOK")
+    public void beforeOKScenario() throws IOException {
+        System.out.println("Setting up mail server");
```
[the remark this answers is not in the document]
Comment (a-breid-ericsson) [reply src/functionaltests/java/com/ericsson/ei/notifications/trigger/SubscriptionNotificationSteps.java]: Already done.

File: src/main/java/com/ericsson/ei/notifications/EmailSender.java [hunk 5]
```diff
@@ -57,64 +58,70 @@
...
+     * This method sends an email.
+     *
+     * @param message   The email message to send
+     * @throws NotificationFailureException
+     */
+    public void sendEmail(MimeMessage message) throws NotificationFailureException {
+        try {
+            emailSender.send(message);
+        } catch (MailException e) {
+            LOGGER.error("", e);
```
[the remark this answers is not in the document]
Comment (a-breid-ericsson) [reply src/main/java/com/ericsson/ei/notifications/EmailSender.java]: Done.

File: src/main/java/com/ericsson/ei/notifications/HttpRequestSender.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            throws AuthenticationException {
+        ResponseEntity<JsonNode> response;
+
+        try {
+            LOGGER.info("Performing HTTP request to url: {}", url);
+            response = rest.postForEntity(url, request, JsonNode.class);
+
+        } catch (HttpClientErrorException e) {
+            checkIfAuthenticationException(e);
+            LOGGER.error("HTTP request failed, bad request! When trying to connect to URL: {}\n{}",
```
Comment (Christoffer-Cortes) [inline src/main/java/com/ericsson/ei/notifications/HttpRequestSender.java]: The last '{}' can be removed, logger will print stack without it.

**Pull request merged**